### DISCUSSION TITLE
Is func 'SetNX' atomic ?

### DISCUSSION THREAD

**Discussion by Wang-Yanping1996** [redis/go-redis#1974] [Q&A]
Is func 'SetNX' atomic ?

```
func (c cmdable) SetNX(key string, value interface{}, expiration time.Duration) *BoolCmd {
	var cmd *BoolCmd
	if expiration == 0 {
		// Use old `SETNX` to support old Redis versions.
		cmd = NewBoolCmd("setnx", key, value)
	} else {
		if usePrecise(expiration) {
			cmd = NewBoolCmd("set", key, value, "px", formatMs(expiration), "nx")
		} else {
			cmd = NewBoolCmd("set", key, value, "ex", formatSec(expiration), "nx")
		}
	}
	_ = c(cmd)
	return cmd
}
```
I think that I cannot use `SETNX` and `EXPIRE` at the same time, but `SET key value ex ... nx` is OK. So the  `SET key value ex ... nx`  is atomic ? The origin code use this format, so the function `SETNX` is atomic ?

I'm not familiar with this, so can you help me ?
Thanks for a lot !

**Comment by vmihailenco** [comment]
All Redis commands are atomic. If you need to execute several command atomically, you can use a pipeline.
